README
This pam module provides the authentication service using an own set of user/password pairs.

CONFIGURATION
=============

simple PAM config
-----------------

Just add/change the config file for service to contain the line:

auth		required	pam_pwdfile.so pwdfile=/path/to/passwd_file

If your service does more with PAM than auth there will be a fallback to the service "other".
If that is not what you want, you can use pam_permit.so or pam_deny.so for that:

account		required	pam_permit.so
session		required	pam_permit.so
password	required	pam_deny.so


options
-------

* pwdfile=<file>
* debug: produce a bit of debug output
* nodelay: don't tell the PAM stack to cause a delay on auth failure
* flock: use a shared (read) advisory lock on pwdfile, you should better move new versions into place instead


PASSWORD FILE
=============

The password file basically looks like passwd(5): one line for each user with two or more colon-separated fields.
First field contains the username, the second the crypt()ed password.
Other fields are optional.

crypt()ed passwords in various formats can be generated with mkpasswd from the whois package.


changelog
commit a7384191e969409740f6412afbfe2f9cdc42944a (HEAD -> master)
Author: supernovatux <thulashitharan.d@gmail.com>
Date:   Thu Apr 6 21:42:00 2023 +0530

    Made the library less secure!!!!!

commit f46bd38f387f0a580e134388086321b03e6b17d3 (origin/master, origin/HEAD)
Author: Timo Weingärtner <timo@tiwe.de>
Date:   Fri Apr 16 19:33:21 2021 +0200

    do away with legacy crypt types
    
    it is the responsibility of libcrypt to implement crypt types

commit e29d26e77d8577ea272211c20240b3c39dff6bf0
Author: Timo Weingärtner <timo@tiwe.de>
Date:   Thu Mar 6 21:30:20 2014 +0100

    enable large file support
    
    this should make some tests for LFS happy; we don't use sizes of or offsets in files here

commit 8f0e412b48178c00abd023917dd2c9050ee89c18 (tag: v1.0)
Author: Timo Weingärtner <timo@tiwe.de>
Date:   Sun Sep 29 15:55:25 2013 +0200

    release 1.0

commit 6946f4bd3102d677d9ce43d4c48ed6bec2b13a31
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-09-29 15:50:21 +0200

    update copyright
    
    also fix spelling error "GNU Public License"

commit 183200793e5aec74f73b21bd96d46b447da64f11
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-09-28 16:35:58 +0200

    Makefile: add target changelog-from-git
    
    needs moreutils

commit 5e5588cd1edfbab1889537b93ae13b159f76caff
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-09-26 16:29:26 +0200

    README: describe legacy_crypt in more detail

commit 235369330e1c1366e5c7cb31561cdb673703ad58
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-09-26 16:29:11 +0200

    fix typo in README

commit 090a8585370c7857a193467fc057828b82ec358f
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-06-01 00:15:00 +0200

    move opening brace out of ifdef to make code folding work

commit 4964a685e75c75d662bb0d4f1bb3fd126f49eca2
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-06-01 00:14:07 +0200

    don't include features.h

commit 201e799f7a2fcba3af2c9214f25545460ef9b08a (tag: v0.100)
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-27 21:10:18 +0200

    rework documentation
    
    remove stuff in contrib, it is outdated or explained in the service's
    documentation

commit da52bf9630a077d90e1338d818a3e179367058c4
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-27 21:08:02 +0200

    separate DESTDIR and PAM_LIB_DIR

commit e493c1467bbaebfbaf2a9a6b1da3398b76232ce5
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-14 20:22:36 +0200

    remove CVS $Id line and static version number

commit caea065f12f3d358948cd0ca760ebd7c27cb6c80
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-14 20:08:06 +0200

    overhaul bigcrypt.c
    
    * drop unnessesary variables
    * rename variables and define's to be more desciptive
    * rotate pointer updates to front of loop
    * don't copy key
    
    there was no point in using crypt_r() here, we return our result in a static
    buffer ourselves

commit 495461432ca4034d49ee37cb398c6bd253d6f66d
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-11 19:34:44 +0200

    md5.c: fix compiler warnings

commit be53f76279d158aa3e5fb2960f9ae4da52201857
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-11 01:43:50 +0200

    replace self-defined uint32 with uint32_t from stdint.h
    
    unsigned int is not guaranteed to have 32 bits

commit ce9367b3202477b3cc914cabfe0cb2a856f3a51d
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-10 21:30:05 +0200

    major overhaul
    
    * merge fgetpwnam into pam_sm_authenticate
    * handle empty password field
    * fix a fd and memory leak if pwdfile opening succeeds but locking fails
    * use crypt_r (enabled via USE_CRYPT_R)
    * rely on crypt() to handle newer crypt variants (including "good" md5 crypt)
    * make bigcrypt and broken md5 crypt optional
    * add some const's

commit 88dd2b1a22cd06fc401a8ddadd41114cebe159d5
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-05-10 21:27:56 +0200

    include proper headers for crypt()
    
    this also prepares for crypt_r()

commit 138c589dd4cdf68659bfa643e5659fa1200f6081
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-29 13:06:26 +0200

    rework pwdfile reading
    
    * drop rewind(), we read the file just once
    * use getline() to get rid of the fixed-size buffer
    * let strsep() also handle the newline
    * stop at the first line containing the user instead of using the last

commit 0437f4656f1d5a541b4ab951c457fae19f8deee4
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-28 17:15:15 +0200

    use pam_get_authtok
    
    this also gets it right with use_first_pass
    also use the default prompt for the username

commit 86c95423b2908869ee42f9f40896a0bb0b773cf4
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-27 23:55:20 +0200

    rework argument parsing
    
    * don't copy pwdfile argument, we don't need to modify it
    * replace sizeof() with strlen() as that is easier to understand and the
      compiler can also optimize it away
    * expand DEFINE's so we can get rid of the comments

commit 64707e82165bb32db5763b38bf550b538bcd4eec
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-27 18:07:22 +0200

    make Brokencrypt_md5 also broken on little-endian
    
    otherwise broken hashes from big-endian systems won't work
    also remove ASM_MD5 #ifndef's, we don't have assembler code here

commit 629c03d7775e1f4b5c0fdee358c6773f70e91961
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-26 12:57:56 +0200

    add debug module option and use pam_syslog
    
    also:
    * remove some unnessesary comments
    * add vim settings for unusual indentation

commit fbce1a480fda4c97b21c87fb39096d23db6eedfb
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-25 14:27:07 +0200

    apply visibility patch by Peter Palfrader

commit 0148de59cdcea4013d694fc04db3174ce06c60b1
Author: Timo Weingärtner <timo@tiwe.de>
Date:   2013-04-25 14:22:49 +0200

    rework Makefile for standalone building
    

pam_pwdfile_plain.c
/* 
 * pam authentication module that can be pointed at any username/crypted
 * text file so that pam using application can use an alternate set of
 * passwords than specified in system password database
 * 
 * Copyright (c) 1999-2003 Charl P. Botha <cpbotha@cpbotha.net>
 * Copyright (c) 2012-2013 Timo Weingärtner <timo@tiwe.de>
 * Copyright (c) 2023 Supernovatux
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 * 
 * ALTERNATIVELY, this product may be distributed under the terms of the
 * GNU General Public License, in which case the provisions of the GPL are
 * required INSTEAD OF the above restrictions.  (This clause is
 * necessary due to a potential bad interaction between the GPL and
 * the restrictions contained in a BSD-style copyright.)
 * 
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <syslog.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/file.h>
#include <unistd.h>
#include <syslog.h>

#include <security/pam_appl.h>

#define PAM_SM_AUTH
#include <security/pam_modules.h>
#include <security/pam_ext.h>

static int lock_fd(int fd) {
    int delay;
    
    for (delay = 5; delay <= 40; delay *= 2) {
	if (flock(fd, LOCK_SH | LOCK_NB) == -1) {
	    // failed
	    if (errno != EWOULDBLOCK) goto failed;
	    sleep(delay);
	}else{
	    return 0;
	}
    }
    if (flock(fd, LOCK_SH | LOCK_NB) != -1) return 0;
    failed:
    return -1;
}
/* expected hook for auth service */
__attribute__((visibility("default")))
PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags,
				   int argc, const char **argv) {
    int i;
    const char *name;
    char const * password;
    char const * pwdfilename = NULL;
    char const * stored_password = NULL;
    FILE *pwdfile;
    int use_flock = 0;
    int use_delay = 1;
    int debug = 1;
    char * linebuf = NULL;
    size_t linebuflen;
    
    /* we require the pwdfile switch and argument to be present, else we don't work */
    for (i = 0; i < argc; ++i) {
	if (!strcmp(argv[i], "pwdfile") && i + 1 < argc)
	    pwdfilename = argv[++i];
	else if (!strncmp(argv[i], "pwdfile=", strlen("pwdfile=")))
	    pwdfilename = argv[i] + strlen("pwdfile=");
	else if (!strcmp(argv[i], "flock"))
	    use_flock = 1;
	else if (!strcmp(argv[i], "noflock"))
	    use_flock = 0;
	else if (!strcmp(argv[i], "nodelay"))
	    use_delay = 0;
	else if (!strcmp(argv[i], "debug"))
	    debug = 1;
    }
    
#ifdef HAVE_PAM_FAIL_DELAY
    if (use_delay) {
	if (debug) pam_syslog(pamh, LOG_DEBUG, "setting fail delay");
	(void) pam_fail_delay(pamh, 2000000);   /* 2 sec */
    }
#endif
    
    if (!pwdfilename) {
	pam_syslog(pamh, LOG_ERR, "password file name not specified");
	return PAM_AUTHINFO_UNAVAIL;
    }
    
    if (pam_get_user(pamh, &name, NULL) != PAM_SUCCESS) {
	pam_syslog(pamh, LOG_ERR, "couldn't get username from PAM stack");
	return PAM_AUTH_ERR;
    }
    if (debug) pam_syslog(pamh, LOG_DEBUG, "username is %s", name);
    
    if (!(pwdfile = fopen(pwdfilename, "r"))) {
	pam_syslog(pamh, LOG_ALERT, "couldn't open password file %s", pwdfilename);
	return PAM_AUTHINFO_UNAVAIL;
    }
    
    if (use_flock && lock_fd(fileno(pwdfile)) == -1) {
	pam_syslog(pamh, LOG_ALERT, "couldn't lock password file %s", pwdfilename);
	fclose(pwdfile);
	return PAM_AUTHINFO_UNAVAIL;
    }
    
    /* get the password corresponding to this user out of pwdfile */
    while (getline(&linebuf, &linebuflen, pwdfile) > 0) {
	/* strsep changes its argument, make a copy */
	char * nexttok = linebuf;
	
	/* first field: username */
	char * curtok = strsep(&nexttok, ":");
	
	/* skip non-matching usernames */
	if (strcmp(curtok, name))
	    continue;
	
	/* second field: password (until next colon or newline) */
	if ((curtok = strsep(&nexttok, ":\n"))) {
	    stored_password = curtok;
	    break;
	}
    }
    fclose(pwdfile);
    /* we keep linebuf (allocated by getline), stored_password is pointing into it */

    if (!stored_password)
	if (debug) pam_syslog(pamh, LOG_ERR, "user not found in password database");
    
    if (stored_password && !strlen(stored_password)) {
	if (debug) pam_syslog(pamh, LOG_DEBUG, "user has empty password field");
	free(linebuf);
	return flags & PAM_DISALLOW_NULL_AUTHTOK ? PAM_AUTH_ERR : PAM_SUCCESS;
    }
    
    if (pam_get_authtok(pamh, PAM_AUTHTOK, &password, NULL) != PAM_SUCCESS) {
	pam_syslog(pamh, LOG_ERR, "couldn't get password from PAM stack");
	free(linebuf);
	return PAM_AUTH_ERR;
    }
    
    if (!stored_password) {
	free(linebuf);
	return PAM_USER_UNKNOWN;
    }
    if (strcmp(password, stored_password)) {
	pam_syslog(pamh, LOG_NOTICE, "wrong password for user %s", name);
	free(linebuf);
	return PAM_AUTH_ERR;
    }
    
    if (debug) pam_syslog(pamh, LOG_DEBUG, "passwords match");
    free(linebuf);
    return PAM_SUCCESS;
}

/* another expected hook */
__attribute__((visibility("default")))
PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, 
			      int argc, const char **argv)
{
    return PAM_SUCCESS;
}

#ifdef PAM_STATIC
struct pam_module _pam_listfile_modstruct = {
    "pam_pwdfile",
	pam_sm_authenticate,
	pam_sm_setcred,
	NULL,
	NULL,
	NULL,
	NULL,
};
#endif


pam_pwdfile.c
/* 
 * pam authentication module that can be pointed at any username/crypted
 * text file so that pam using application can use an alternate set of
 * passwords than specified in system password database
 * 
 * Copyright (c) 1999-2003 Charl P. Botha <cpbotha@cpbotha.net>
 * Copyright (c) 2012-2013 Timo Weingärtner <timo@tiwe.de>
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 * 
 * ALTERNATIVELY, this product may be distributed under the terms of the
 * GNU General Public License, in which case the provisions of the GPL are
 * required INSTEAD OF the above restrictions.  (This clause is
 * necessary due to a potential bad interaction between the GPL and
 * the restrictions contained in a BSD-style copyright.)
 * 
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifdef USE_CRYPT_R
#define _GNU_SOURCE
#include <crypt.h>
#else
#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 700
#endif
#ifndef _BSD_SOURCE
#define _BSD_SOURCE
#endif
#endif

#include <syslog.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/file.h>
#include <unistd.h>
#include <syslog.h>

#include <security/pam_appl.h>

#define PAM_SM_AUTH
#include <security/pam_modules.h>
#include <security/pam_ext.h>

static int lock_fd(int fd) {
    int delay;
    
    for (delay = 5; delay <= 40; delay *= 2) {
	if (flock(fd, LOCK_SH | LOCK_NB) == -1) {
	    /* failed */
	    if (errno != EWOULDBLOCK) goto failed;
	    sleep(delay);
	}else{
	    return 0;
	}
    }
    if (flock(fd, LOCK_SH | LOCK_NB) != -1) return 0;
    failed:
    return -1;
}

/* expected hook for auth service */
__attribute__((visibility("default")))
PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags,
				   int argc, const char **argv) {
    int i;
    const char *name;
    char const * password;
    char const * pwdfilename = NULL;
    char const * stored_crypted_password = NULL;
    char const * crypted_password;
    FILE *pwdfile;
    int use_flock = 0;
    int use_delay = 1;
    int debug = 0;
    char * linebuf = NULL;
    size_t linebuflen;
#ifdef USE_CRYPT_R
    struct crypt_data crypt_buf;
#endif
    
    /* we require the pwdfile switch and argument to be present, else we don't work */
    for (i = 0; i < argc; ++i) {
	if (!strcmp(argv[i], "pwdfile") && i + 1 < argc)
	    pwdfilename = argv[++i];
	else if (!strncmp(argv[i], "pwdfile=", strlen("pwdfile=")))
	    pwdfilename = argv[i] + strlen("pwdfile=");
	else if (!strcmp(argv[i], "flock"))
	    use_flock = 1;
	else if (!strcmp(argv[i], "noflock"))
	    use_flock = 0;
	else if (!strcmp(argv[i], "nodelay"))
	    use_delay = 0;
	else if (!strcmp(argv[i], "debug"))
	    debug = 1;
    }
    
#ifdef HAVE_PAM_FAIL_DELAY
    if (use_delay) {
	if (debug) pam_syslog(pamh, LOG_DEBUG, "setting fail delay");
	(void) pam_fail_delay(pamh, 2000000);   /* 2 sec */
    }
#endif
    
    if (!pwdfilename) {
	pam_syslog(pamh, LOG_ERR, "password file name not specified");
	return PAM_AUTHINFO_UNAVAIL;
    }
    
    if (pam_get_user(pamh, &name, NULL) != PAM_SUCCESS) {
	pam_syslog(pamh, LOG_ERR, "couldn't get username from PAM stack");
	return PAM_AUTH_ERR;
    }
    if (debug) pam_syslog(pamh, LOG_DEBUG, "username is %s", name);
    
    if (!(pwdfile = fopen(pwdfilename, "r"))) {
	pam_syslog(pamh, LOG_ALERT, "couldn't open password file %s", pwdfilename);
	return PAM_AUTHINFO_UNAVAIL;
    }
    
    if (use_flock && lock_fd(fileno(pwdfile)) == -1) {
	pam_syslog(pamh, LOG_ALERT, "couldn't lock password file %s", pwdfilename);
	fclose(pwdfile);
	return PAM_AUTHINFO_UNAVAIL;
    }
    
    /* get the crypted password corresponding to this user out of pwdfile */
    while (getline(&linebuf, &linebuflen, pwdfile) > 0) {
	/* strsep changes its argument, make a copy */
	char * nexttok = linebuf;
	
	/* first field: username */
	char * curtok = strsep(&nexttok, ":");
	
	/* skip non-matching usernames */
	if (strcmp(curtok, name))
	    continue;
	
	/* second field: password (until next colon or newline) */
	if ((curtok = strsep(&nexttok, ":\n"))) {
	    stored_crypted_password = curtok;
	    break;
	}
    }
    fclose(pwdfile);
    /* we keep linebuf (allocated by getline), stored_crypted_password is pointing into it */

    if (!stored_crypted_password)
	if (debug) pam_syslog(pamh, LOG_ERR, "user not found in password database");
    
    if (stored_crypted_password && !strlen(stored_crypted_password)) {
	if (debug) pam_syslog(pamh, LOG_DEBUG, "user has empty password field");
	free(linebuf);
	return flags & PAM_DISALLOW_NULL_AUTHTOK ? PAM_AUTH_ERR : PAM_SUCCESS;
    }
    
    if (pam_get_authtok(pamh, PAM_AUTHTOK, &password, NULL) != PAM_SUCCESS) {
	pam_syslog(pamh, LOG_ERR, "couldn't get password from PAM stack");
	free(linebuf);
	return PAM_AUTH_ERR;
    }
    
    if (!stored_crypted_password) {
	free(linebuf);
	return PAM_USER_UNKNOWN;
    }
    
    if (debug) pam_syslog(pamh, LOG_DEBUG, "got crypted password == '%s'", stored_crypted_password);
    
#ifdef USE_CRYPT_R
    crypt_buf.initialized = 0;
    if (!(crypted_password = crypt_r(password, stored_crypted_password, &crypt_buf)))
#else
    if (!(crypted_password = crypt(password, stored_crypted_password)))
#endif
    {
	pam_syslog(pamh, LOG_ERR, "crypt() failed");
	free(linebuf);
	return PAM_AUTH_ERR;
    }
    
    if (strcmp(crypted_password, stored_crypted_password)) {
	pam_syslog(pamh, LOG_NOTICE, "wrong password for user %s", name);
	free(linebuf);
	return PAM_AUTH_ERR;
    }
    
    if (debug) pam_syslog(pamh, LOG_DEBUG, "passwords match");
    free(linebuf);
    return PAM_SUCCESS;
}

/* another expected hook */
__attribute__((visibility("default")))
PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, 
			      int argc, const char **argv)
{
    return PAM_SUCCESS;
}

#ifdef PAM_STATIC
struct pam_module _pam_listfile_modstruct = {
    "pam_pwdfile",
	pam_sm_authenticate,
	pam_sm_setcred,
	NULL,
	NULL,
	NULL,
	NULL,
};
#endif
/* vim:set ts=8 sw=4: */
